File: platform/macos/Clipboard/NativeClipboardCoordinator.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation

@MainActor public final class NativeClipboardCoordinator {
  private final class Registration {
    weak var session: NativeSession?
    let report: (String?) -> Void
    var subscriptions = Set<AnyCancellable>()
    init(_ session: NativeSession, _ report: @escaping (String?) -> Void) { self.session = session; self.report = report }
  }
  private struct RemoteJob {
    let session: NativeSession
    let update: NativeClipboardUpdate
    let epoch: UInt64
  }
  private let pasteboard: any NativePasteboardAccess
  private let automaticPolling: Bool
  private var registrations: [ObjectIdentifier: Registration] = [:]
  private weak var active: NativeSession?
  private var applicationActive = true
  private var stopped = false, reconciliationQueued = false
  private var polling: Task<Void, Never>?
  private var epoch: UInt64 = 0
  private var observedChange: Int?
  private var sendEnabled = false
  private var pendingRemote: RemoteJob?
  private var pollRequested = false
  private var operation: Task<Void, Never>?
  public init(pasteboard: any NativePasteboardAccess = NativePasteboard(), automaticPolling: Bool = true) {
    self.pasteboard = pasteboard; self.automaticPolling = automaticPolling
  }
  deinit { polling?.cancel(); operation?.cancel() }
  public func register(_ session: NativeSession, onStatus: @escaping (String?) -> Void = { _ in }) {
    guard !stopped, registrations[ObjectIdentifier(session)] == nil else { return }
    let entry = Registration(session, onStatus); registrations[ObjectIdentifier(session)] = entry
    let changes = [session.$isFocused.removeDuplicates().map { _ in () }.eraseToAnyPublisher(),
      session.$isClosing.removeDuplicates().map { _ in () }.eraseToAnyPublisher(),
      session.$isViewOnly.removeDuplicates().map { _ in () }.eraseToAnyPublisher(),
      session.$snapshot.map(\.state).removeDuplicates().map { _ in () }.eraseToAnyPublisher(),
      session.$clipboardSendEnabled.removeDuplicates().map { _ in () }.eraseToAnyPublisher(),
      session.$clipboardReceiveEnabled.removeDuplicates().map { _ in () }.eraseToAnyPublisher()]
    Publishers.MergeMany(changes).sink { [weak self] in MainActor.assumeIsolated { self?.routingChanged() } }.store(in: &entry.subscriptions)
    // Registration must not replay a cached remote copy over a newer OS copy.
    session.$clipboard.dropFirst().sink { [weak self, weak session] update in MainActor.assumeIsolated {
      if let session, let update { self?.receive(update, from: session) }
    } }.store(in: &entry.subscriptions)
    scheduleReconciliation()
  }
  public func unregister(_ session: NativeSession) {
    registrations.removeValue(forKey: ObjectIdentifier(session)); reconcile()
  }
  public func setApplicationActive(_ enabled: Bool) {
    applicationActive = enabled
    if !enabled {
      // Also invalidates already-admitted sends and delayed remote replies.
      for entry in registrations.values { try? entry.session?.setFocused(false) }
    }
    reconcile()
  }
  public func stop() {
    stopped = true; polling?.cancel(); polling = nil
    pendingRemote = nil; pollRequested = false; operation?.cancel(); active = nil; registrations.removeAll()
  }
  // Cancellation skips queued native access. An already executing OS call cannot
  // be interrupted; close drains it asynchronously and discards its result.
  public func close() async { stop(); await operation?.value }
  private func scheduleReconciliation() {
    guard !stopped, !reconciliationQueued else { return }
    reconciliationQueued = true
    // Published emits before storing its value. Reconcile next actor turn and
    // coalesce focus/state changes, rather than reading the old property value.
    Task { @MainActor [weak self] in
      guard let self else { return }
      self.reconciliationQueued = false; self.reconcile()
    }
  }
  private func routingChanged() {
    // Invalidate immediately even if focus loses/regains or policy toggles twice
    // before the coalesced reconciliation runs. Final-state equality is not proof
    // that pending clipboard work still belongs to the same focus interval.
    epoch &+= 1; pendingRemote = nil; pollRequested = false; operation?.cancel(); observedChange = nil
    scheduleReconciliation()
  }
  private func eligible(_ session: NativeSession) -> Bool {
    applicationActive && !stopped && session.isFocused && !session.isViewOnly && !session.isClosing && session.snapshot.state == .connected
  }
  private func reconcile() {
    guard !stopped else { return }
    registrations = registrations.filter { $0.value.session != nil }
    let candidates = registrations.values.compactMap(\.session).filter { eligible($0) }
    // Ambiguous host focus never routes clipboard data to an arbitrary session.
    let selected = candidates.count == 1 ? candidates[0] : nil
    let enabled = selected?.clipboardSendEnabled == true
    if active !== selected || sendEnabled != enabled {
      active = selected; sendEnabled = enabled; epoch &+= 1; observedChange = nil
      pendingRemote = nil; pollRequested = false; operation?.cancel()
    }
    if active != nil && sendEnabled && automaticPolling && polling == nil {
      polling = Task { @MainActor [weak self] in
        while !Task.isCancelled {
          do { try await Task.sleep(for: .milliseconds(250)) } catch { break }
          self?.poll()
        }
      }
    } else if active == nil || !sendEnabled { polling?.cancel(); polling = nil }
    pollCurrent()
  }
  // Exposed for an event-driven host and deterministic adapter tests. The app
  // uses one cancellable 250 ms observation task while a desktop can send.
  public func poll() { reconcile() }
  private func pollCurrent() {
    guard let session = active, eligible(session), sendEnabled else { return }
    pollRequested = true
    startPending()
  }
  private func current(_ session: NativeSession, generation: UInt64, epoch: UInt64) -> Bool {
    !Task.isCancelled && active === session && eligible(session) && sendEnabled &&
      self.epoch == epoch && generation == session.generation
  }
  private func startPending() {
    // One operation owns all native access, including reads and remote writes.
    // Poll ticks collapse to a bit and remote updates to the latest value. Never
    // launch another read just because the previous OS request was cancelled.
    guard !stopped, operation == nil, pendingRemote != nil || pollRequested else { return }
    operation = Task { @MainActor [weak self] in
      guard let self else { return }
      if let remote = self.pendingRemote {
        self.pendingRemote = nil
        await self.writeRemote(remote)
      } else {
        self.pollRequested = false
        if let session = self.active {
          await self.readLocal(session, generation: session.generation, epoch: self.epoch)
        }
      }
      self.operation = nil
      self.startPending()
    }
  }
  private func readLocal(_ session: NativeSession, generation: UInt64, epoch: UInt64) async {
    guard current(session, generation: generation, epoch: epoch) else { return }
    var sampledChange: Int?
    do {
      let change = try await pasteboard.currentChange()
      guard current(session, generation: generation, epoch: epoch), observedChange != change else { return }
      sampledChange = change
      let content = try await pasteboard.read(expectedChange: change, maximumBytes: 256 * 1024)
      guard current(session, generation: generation, epoch: epoch) else { return }
      let latest = try await pasteboard.currentChange()
      guard current(session, generation: generation, epoch: epoch) else { return }
      guard latest == change else { pollRequested = true; return }
      observedChange = change
      switch content {
      case .text(let text):
        _ = try await session.offerClipboard(text, expectedGeneration: generation)
        if current(session, generation: generation, epoch: epoch) { report(session, nil) }
      case .remote, .unavailable:
        _ = try await session.clearClipboard(expectedGeneration: generation)
      }
    } catch is CancellationError {
    } catch NativePasteboardError.changed {
      // Retry on the next tick rather than spinning against an unstable owner.
    } catch let error as NativeError where [.stale, .notConnected, .closing, .unfocused, .viewOnly, .disabled, .echo].contains(error.status) {
    } catch {
      guard current(session, generation: generation, epoch: epoch) else { return }
      // Failures from an obsolete clipboard snapshot must not clear a newer offer
      // or display an error for a connection that no longer owns the operation.
      guard let change = sampledChange,
            let latest = try? await pasteboard.currentChange(), latest == change,
            current(session, generation: generation, epoch: epoch) else { return }
      observedChange = change
      if error is NativePasteboardError {
        report(session, String(localized:"clipboard.recovery.the.local.clipboard.could.not.be.sent.copy.plain.text.of.at", defaultValue:"The local clipboard could not be sent. Copy plain text of at most 256 KiB and try again."))
        _ = try? await session.clearClipboard(expectedGeneration: generation)
      } else {
        report(session, String(localized:"clipboard.recovery.clipboard.transfer.failed.copy.the.text.again.to.retry", defaultValue:"Clipboard transfer failed. Copy the text again to retry."))
      }
    }
  }
  private func receive(_ update: NativeClipboardUpdate, from session: NativeSession) {
    guard !stopped else { return }
    let candidates = registrations.values.compactMap(\.session).filter { eligible($0) }
    guard candidates.count == 1, candidates.first === session else { return }
    if update.kind == .rejected { report(session, String(localized:"clipboard.recovery.the.remote.clipboard.text.could.not.be.accepted", defaultValue:"The remote clipboard text could not be accepted.")); return }
    guard update.kind == .text, update.text != nil else { return }
    do { try session.validateClipboard(update.route, sending: false) } catch { return }
    epoch &+= 1
    operation?.cancel()
    pollRequested = false
    pendingRemote = RemoteJob(session: session, update: update, epoch: epoch)
    startPending()
  }
  private func writeRemote(_ job: RemoteJob) async {
    let session = job.session
    guard !stopped, !Task.isCancelled, epoch == job.epoch, eligible(session),
          let text = job.update.text else { return }
    do {
      try session.validateClipboard(job.update.route, sending: false)
      let change = try await pasteboard.writeRemote(text.text, maximumBytes: 256 * 1024)
      guard !stopped, !Task.isCancelled, epoch == job.epoch, eligible(session) else { return }
      try session.validateClipboard(job.update.route, sending: false)
      observedChange = change
      report(session, nil)
    } catch is CancellationError {
    } catch NativePasteboardError.changed { observedChange = nil }
    catch let error as NativeError where [.stale, .notConnected, .closing, .unfocused, .viewOnly, .disabled].contains(error.status) {}
    catch {
      if !stopped, !Task.isCancelled, epoch == job.epoch, eligible(session) {
        report(session, String(localized:"clipboard.recovery.the.remote.clipboard.could.not.be.written.on.this.mac.copy.it", defaultValue:"The remote clipboard could not be written on this Mac. Copy it again to retry."))
      }
    }
  }
  private func report(_ session: NativeSession, _ message: String?) { registrations[ObjectIdentifier(session)]?.report(message) }
}

File: platform/windows/TidyVNC.Native/Clipboard/NativeClipboardCoordinator.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.ComponentModel;

namespace TidyVNC.Native.Clipboard;

/// <summary>A clipboard problem shown on the affected connection (the app localizes it).</summary>
public enum NativeClipboardNotice
{
    /// <summary>The local clipboard could not be sent (not plain text of at most 256 KiB, or unavailable).</summary>
    LocalNotSent,
    /// <summary>The transfer failed; copying again retries.</summary>
    TransferFailed,
    /// <summary>The remote clipboard text could not be accepted.</summary>
    RemoteRejected,
    /// <summary>The remote clipboard could not be written on this PC.</summary>
    RemoteNotWritten,
}

/// <summary>
/// The app-wide clipboard router (SERVICES.md section 6; NativeClipboardCoordinator
/// on macOS). Text goes only to the one connection whose desktop is focused
/// in the active app, is connected and not view-only; ambiguous focus routes
/// nothing. A local change is offered when it happens (clipboard listener)
/// and again when focus returns, if it changed meanwhile. Remote text is
/// written with provenance, and text carrying this process's provenance is
/// never sent back. One operation owns all clipboard access at a time; any
/// routing change invalidates pending work. UI thread only.
/// </summary>
public sealed class NativeClipboardCoordinator : IDisposable
{
    public const int MaximumBytes = 256 * 1024;

    private sealed class Registration(NativeSession session, Action<NativeClipboardNotice?> report)
    {
        public NativeSession Session { get; } = session;
        public Action<NativeClipboardNotice?> Report { get; } = report;
        public PropertyChangedEventHandler? Handler { get; set; }
        /// <summary>Only state transitions route; frame and counter updates do not.</summary>
        public NativeSessionState State { get; set; } = session.Snapshot.State;
    }

    private sealed record RemoteJob(NativeSession Session, NativeClipboardUpdate Update, ulong Epoch);

    private static readonly string[] RoutingProperties =
    [
        nameof(NativeSession.IsFocused), nameof(NativeSession.IsClosing), nameof(NativeSession.IsViewOnly),
        nameof(NativeSession.ClipboardSendEnabled), nameof(NativeSession.ClipboardReceiveEnabled),
    ];

    private readonly IUiDispatcher dispatcher;
    private readonly INativeClipboardAccess clipboard;
    private readonly List<Registration> registrations = [];
    private NativeSession? active;
    private bool applicationActive = true, stopped, reconciliationQueued, sendEnabled, pollRequested;
    private ulong epoch;
    private uint? observedChange;
    private RemoteJob? pendingRemote;
    private Task? operation;
    private CancellationTokenSource? operationCancel;
    private readonly Action changed;

    public NativeClipboardCoordinator(IUiDispatcher dispatcher, INativeClipboardAccess clipboard)
    {
        this.dispatcher = dispatcher;
        this.clipboard = clipboard;
        changed = () => dispatcher.TryEnqueue(Poll);
        clipboard.Changed += changed;
    }

    /// <summary>The operation in flight (tests and close await it).</summary>
    public Task Operation => operation ?? Task.CompletedTask;

    public void Register(NativeSession session, Action<NativeClipboardNotice?>? onStatus = null)
    {
        UiThread.Require(dispatcher);
        if (stopped || registrations.Any(r => ReferenceEquals(r.Session, session))) return;
        var entry = new Registration(session, onStatus ?? (_ => { }));
        entry.Handler = (_, change) =>
        {
            if (change.PropertyName is { } name && RoutingProperties.Contains(name)) RoutingChanged();
            else if (change.PropertyName == nameof(NativeSession.Snapshot))
            {
                if (session.Snapshot.State == entry.State) return;
                entry.State = session.Snapshot.State;
                RoutingChanged();
            }
            // Registration never replays a cached remote copy over a newer local one.
            else if (change.PropertyName == nameof(NativeSession.Clipboard) && session.Clipboard is { } update) Receive(update, session);
        };
        session.PropertyChanged += entry.Handler;
        registrations.Add(entry);
        ScheduleReconciliation();
    }

    public void Unregister(NativeSession session)
    {
        UiThread.Require(dispatcher);
        foreach (var entry in registrations.Where(r => ReferenceEquals(r.Session, session)).ToList())
        {
            session.PropertyChanged -= entry.Handler;
            registrations.Remove(entry);
        }
        Reconcile();
    }

    /// <summary>App activation: losing it unfocuses every desktop, which also invalidates admitted sends.</summary>
    public void SetApplicationActive(bool enabled)
    {
        UiThread.Require(dispatcher);
        applicationActive = enabled;
        if (!enabled)
            foreach (var entry in registrations)
                try { if (entry.Session.IsFocused) entry.Session.SetFocused(false); }
                catch (NativeError) { }
        Reconcile();
    }

    private void CancelOperation()
    {
        operationCancel?.Cancel();
    }

    public void Stop()
    {
        stopped = true;
        clipboard.Changed -= changed;
        pendingRemote = null;
        pollRequested = false;
        CancelOperation();
        active = null;
        foreach (var entry in registrations) entry.Session.PropertyChanged -= entry.Handler;
        registrations.Clear();
    }

    /// <summary>An executing clipboard call cannot be interrupted; close waits for it and discards its result.</summary>
    public async Task CloseAsync()
    {
        Stop();
        await Operation;
    }

    public void Dispose() => Stop();

    private void ScheduleReconciliation()
    {
        if (stopped || reconciliationQueued) return;
        reconciliationQueued = true;
        // Coalesce focus/state changes into one reconciliation on the next turn.
        dispatcher.TryEnqueue(() =>
        {
            reconciliationQueued = false;
            Reconcile();
        });
    }

    private void RoutingChanged()
    {
        // Final-state equality is not proof that pending work still belongs to the same focus interval.
        epoch++;
        pendingRemote = null;
        pollRequested = false;
        CancelOperation();
        observedChange = null;
        ScheduleReconciliation();
    }

    private bool Eligible(NativeSession session)
        => applicationActive && !stopped && session.IsFocused && !session.IsViewOnly && !session.IsClosing &&
           session.Snapshot.State == NativeSessionState.Connected;

    private void Reconcile()
    {
        if (stopped) return;
        var candidates = registrations.Select(r => r.Session).Where(Eligible).ToList();
        // Ambiguous focus never routes clipboard data to an arbitrary session.
        var selected = candidates.Count == 1 ? candidates[0] : null;
        var enabled = selected?.ClipboardSendEnabled == true;
        if (!ReferenceEquals(active, selected) || sendEnabled != enabled)
        {
            active = selected;
            sendEnabled = enabled;
            epoch++;
            observedChange = null;
            pendingRemote = null;
            pollRequested = false;
            CancelOperation();
        }
        PollCurrent();
    }

    /// <summary>A clipboard change notification (or a test's explicit check).</summary>
    public void Poll()
    {
        if (dispatcher.HasThreadAccess) Reconcile();
    }

    /// <summary>App-originated copies (e.g. redacted diagnostics); routing then treats them like any local copy.</summary>
    public async Task CopyLocalAsync(string text)
    {
        UiThread.Require(dispatcher);
        if (stopped) throw new NativeClipboardAccessException(NativeClipboardAccessError.Closed);
        await clipboard.WriteLocalAsync(text, 1 << 20);
        Reconcile();
    }

    private void PollCurrent()
    {
        if (active is not { } session || !Eligible(session) || !sendEnabled) return;
        pollRequested = true;
        StartPending();
    }

    private bool Current(NativeSession session, ulong generation, ulong ticket, CancellationToken token)
        => !token.IsCancellationRequested && ReferenceEquals(active, session) && Eligible(session) && sendEnabled &&
           epoch == ticket && generation == session.Generation;

    private void StartPending()
    {
        // One operation owns all clipboard access; polls collapse to a flag and remote updates to the latest value.
        if (stopped || operation is not null || (pendingRemote is null && !pollRequested)) return;
        var cancel = new CancellationTokenSource();
        operationCancel = cancel;
        operation = Run();

        async Task Run()
        {
            try
            {
                await Task.Yield();
                // Cancelled before it began: leave the pending work to a fresh operation.
                if (cancel.IsCancellationRequested) return;
                if (pendingRemote is { } remote)
                {
                    pendingRemote = null;
                    await WriteRemote(remote, cancel.Token);
                }
                else
                {
                    pollRequested = false;
                    if (active is { } session) await ReadLocal(session, session.Generation, epoch, cancel.Token);
                }
            }
            finally
            {
                if (ReferenceEquals(operationCancel, cancel)) operationCancel = null;
                cancel.Dispose();
                operation = null;
                StartPending();
            }
        }
    }

    private static bool Quiet(Exception error)
        => error is NativeError { Status: NativeStatus.Stale or NativeStatus.NotConnected or NativeStatus.Closing or NativeStatus.Unfocused
                                  or NativeStatus.ViewOnly or NativeStatus.Disabled or NativeStatus.Echo }
           || error is NativeCommandFailure { Result: NativeCommandFailure.ResultKind.Cancelled };

    private async Task ReadLocal(NativeSession session, ulong generation, ulong ticket, CancellationToken token)
    {
        if (!Current(session, generation, ticket, token)) return;
        uint? sampled = null;
        try
        {
            var change = await clipboard.CurrentChangeAsync();
            if (!Current(session, generation, ticket, token) || observedChange == change) return;
            sampled = change;
            var content = await clipboard.ReadAsync(change, MaximumBytes);
            if (!Current(session, generation, ticket, token)) return;
            var latest = await clipboard.CurrentChangeAsync();
            if (!Current(session, generation, ticket, token)) return;
            if (latest != change) { pollRequested = true; return; }
            observedChange = change;
            if (content is { Kind: NativeClipboardContentKind.Text, Text: { } text })
            {
                await session.OfferClipboardAsync(text, changeId: change, expectedGeneration: generation);
                if (Current(session, generation, ticket, token)) Report(session, null);
            }
            else await session.ClearClipboardAsync(generation);
        }
        catch (NativeClipboardAccessException error) when (error.Error == NativeClipboardAccessError.Changed)
        {
            // Retried on the next change rather than spinning against an unstable owner.
        }
        catch (Exception error) when (Quiet(error)) { }
        catch (Exception error) when (error is NativeClipboardAccessException or NativeError or NativeCommandFailure)
        {
            if (!Current(session, generation, ticket, token)) return;
            // A failure from an obsolete snapshot must not clear a newer offer or report on another connection.
            if (sampled is not { } change) return;
            uint latest;
            try { latest = await clipboard.CurrentChangeAsync(); }
            catch (NativeClipboardAccessException) { return; }
            if (latest != change || !Current(session, generation, ticket, token)) return;
            observedChange = change;
            if (error is NativeClipboardAccessException)
            {
                Report(session, NativeClipboardNotice.LocalNotSent);
                try { await session.ClearClipboardAsync(generation); }
                catch (Exception clear) when (clear is NativeError or NativeCommandFailure) { }
            }
            else Report(session, NativeClipboardNotice.TransferFailed);
        }
    }

    private void Receive(NativeClipboardUpdate update, NativeSession session)
    {
        if (stopped) return;
        var candidates = registrations.Select(r => r.Session).Where(Eligible).ToList();
        if (candidates.Count != 1 || !ReferenceEquals(candidates[0], session)) return;
        if (update.Kind == NativeClipboardUpdate.UpdateKind.Rejected) { Report(session, NativeClipboardNotice.RemoteRejected); return; }
        if (update.Kind != NativeClipboardUpdate.UpdateKind.Text || update.Text is null) return;
        try { session.ValidateClipboard(update.Route, sending: false); }
        catch (NativeError) { return; }
        // Settle routing first, so a queued reconciliation for this same focus
        // change cannot discard text that arrived just before it ran.
        if (!ReferenceEquals(active, session)) Reconcile();
        epoch++;
        CancelOperation();
        pollRequested = false;
        pendingRemote = new RemoteJob(session, update, epoch);
        StartPending();
    }

    private async Task WriteRemote(RemoteJob job, CancellationToken token)
    {
        var session = job.Session;
        if (stopped || token.IsCancellationRequested || epoch != job.Epoch || !Eligible(session) || job.Update.Text is not { } text) return;
        try
        {
            session.ValidateClipboard(job.Update.Route, sending: false);
            var change = await clipboard.WriteRemoteAsync(text.Text, job.Update.Route.SessionIdentity, job.Update.Route.Generation, MaximumBytes);
            if (stopped || token.IsCancellationRequested || epoch != job.Epoch || !Eligible(session)) return;
            session.ValidateClipboard(job.Update.Route, sending: false);
            observedChange = change;
            Report(session, null);
        }
        catch (NativeClipboardAccessException error) when (error.Error == NativeClipboardAccessError.Changed) { observedChange = null; }
        catch (NativeError error) when (Quiet(error)) { }
        catch (Exception error) when (error is NativeClipboardAccessException or NativeError)
        {
            if (!stopped && !token.IsCancellationRequested && epoch == job.Epoch && Eligible(session)) Report(session, NativeClipboardNotice.RemoteNotWritten);
        }
    }

    private void Report(NativeSession session, NativeClipboardNotice? notice)
        => registrations.FirstOrDefault(r => ReferenceEquals(r.Session, session))?.Report(notice);
}
